### ISSUE TITLE
Binary target dir name has a "-" instead of "_"

### ISSUE THREAD

**Issue by pfeairheller** [constabulary/gb#634]
At line 300 of gb/context.go, gb is creating the target dir name based on architecture and os by concatenating them with a "-" instead of the (seemingly?) standard "_".  Here is the relevant function:

`// ctxString returns a string representation of the unique properties
// of the context.
func (c *Context) ctxString() string {
	v := []string{
		c.gotargetos,
		c.gotargetarch,
	}
	v = append(v, c.buildtags...)
	return strings.Join(v, "-")
}
`

This is causing issue with our code generators that use the "types" package and the default importer with the following:

`	conf := types.Config{Importer: importer.Default()}
`

The default importer uses built binaries to discover dependencies and since they are in a pkg directory (darwin-amd64 in our case) with the "-" instead of the "_", it can't find any of our code.

Was the use of the "-" intentional?  Would it be possible to change it to an "_" to match the what the go/importer lib expects?

Thank you.

**Comment by davecheney** [comment]
Yes, the `-` is intentional to avoid overwriting .a files created by the `go` tool.

**Comment by gandalfSievers** [comment]
This causes problems when using `go vet`. It needs the imports as binaries in order to perform all checks. At the moment `go vet` just silently ignores the missing binaries and exits with `0`, if you don't use `-v` to find that it says something like
```
src/foo/bar.go:7:2: could not import baz/foz (can't find import: "baz/foz")
```

While it would find an error like:

```
src/foo/bar.go:191: baz/foz.MyType composite literal uses unkeyed fields
```
and exit with `1` if it can discover the binaries.

**Comment by paraboul** [comment]
GoCode is using generated .a files for completion.

https://github.com/constabulary/gb/issues/634

```
[09:34:30] margo: agent#010: log: Gocode env: gocode.margoEnv{LibPath:"/usr/lib/go/pkg/linux_amd64:/home/paraboul/go/src/git.xxx.com/x.yyy.com/api-server/vendor/pkg/linux_amd64:/home/paraboul/go/src/git.xxx.com/x.yyy.com/api-server/pkg/linux_amd64", GOOS:"linux", GOARCH:"amd64", Compiler:"gc", GOROOT:"/usr/lib/go", GOPATH:"/home/paraboul/go/src/git.xxx.com/x.yyy.com/api-server/vendor:/home/paraboul/go/src/git.xxx.com/x.yyy.com/api-server", InstallSuffix:""}
```
`linux_amd64` vs `linux-amd64`

Haven't found a proper fix yet...

**Comment by davecheney** [comment]
This is not an answer, but an explanation. gb creates .a files in a slightly different directory layout because I was not sure if they would be compatible with output from the go tool.

**Comment by paraboul** [comment]
@davecheney My post was actually for another project referencing this issue :)
